app/models/scheme.rb

```diff
@@ -7,103 +9,139 @@ class Scheme < ApplicationRecord
+
+  def check_details_attributes
     [
       { name: "Service code", value: code },
       { name: "Name", value: service_name },
-      { name: "Confidential information", value: sensitive_display },
+      { name: "Confidential information", value: sensitive },
       { name: "Managed by", value: organisation.name },
-      { name: "Type of scheme", value: scheme_type_display },
-      { name: "Registered under Care Standards Act 2000", value: registered_under_care_act_display },
+      { name: "Type of scheme", value: scheme_type },
+      { name: "Registered under Care Standards Act 2000", value: registered_under_care_act },
       { name: "Total number of units", value: total_units },
-      { name: "Primary client group", value: primary_client_group_display },
-      { name: "Secondary client group", value: secondary_client_group_display },
-      { name: "Level of support given", value: support_type_display },
-      { name: "Intended length of stay", value: intended_stay_display },
     ]
   end
 
-  def scheme_type_display
-    SCHEME_TYPE[scheme_type]
+  def check_primary_client_attributes
+    [
+      { name: "Primary client group", value: primary_client_group },
+    ]
   end
 
-  def sensitive_display
-    SENSITIVE[sensitive]
+  def check_secondary_client_confirmation_attributes
+    [
+      { name: "Has another client group", value: has_other_client_group },
+    ]
   end
 
-  def registered_under_care_act_display
-    REGISTERED_UNDER_CARE_ACT[registered_under_care_act]
+  def check_secondary_client_attributes
+    [
+      { name: "Secondary client group", value: secondary_client_group },
+    ]
   end
 
-  def primary_client_group_display
-    PRIMARY_CLIENT_GROUP[primary_client_group]
+  def check_support_attributes
+    [
+      { name: "Level of support given", value: support_type },
+      { name: "Intended length of stay", value: intended_stay },
+    ]
   end
 
-  def secondary_client_group_display
-    PRIMARY_CLIENT_GROUP[secondary_client_group]
+  def display_attributes
+    [
+      { name: "Service code", value: code },
+      { name: "Name", value: service_name },
+      { name: "Confidential information", value: sensitive },
+      { name: "Managed by", value: organisation.name },
+      { name: "Type of scheme", value: scheme_type },
+      { name: "Registered under Care Standards Act 2000", value: registered_under_care_act },
+      { name: "Total number of units", value: total_units },
+      { name: "Primary client group", value: primary_client_group },
+      { name: "Secondary client group", value: secondary_client_group },
+      { name: "Level of support given", value: support_type },
+      { name: "Intended length of stay", value: intended_stay },
+    ]
   end
 
-  def support_type_display
-    SUPPORT_TYPE[support_type]
-  end
+private
 
-  def intended_stay_display
-    INTENDED_STAY[intended_stay]
+  def create_code
+    self.code = Scheme.last.nil? ? "S1" : "S#{Scheme.last.code[1..].to_i + 1}"
```

Comment:
Is this field (and the `before_create`) actually needed at all or could we just prepend "S" to "scheme.id" where ever we display it? It doesn't seem worth storing?

Reply:
Also "Scheme.last.code" seems unlikely to be race condition safe

Reply:
Then we need to write logic not to prefix S for the imported schemes and also I am not sure that import not going to bring repeated codes in, this makes me think that potentially there is a better solution to address 'code' issue. Noted in the non race condition safe, will look into that.

Reply:
Might be worth syncing up on but my expectation is that old schemes are imported with an auto-incrementing "id" (same as users or case logs), and we also save an additional "old_id" or "old_code" field that has the old value (and will be unique since it uniquely identified the scheme in the old system - new schemes won't have a value for this field).

On display we "ideally" always display the auto-incrementing "id" with "S" prepended. If we need to display the old code, we can do that if it's set _or_ display the id if it isn't. i.e. something like 

```ruby
def display_code
    scheme.old_code || "S" + scheme.id.to_s
end
```